src/layers/types.jl

```diff
@@ -0,0 +1,91 @@
+import Adapt, Functors, Optimisers
+
+"""
+    Flux.AbstractLayer
+    
+Supertype for all of Flux's built-in layers.
+
+Layer types are not essential to use your own `struct` with Flux.
+But they do simplify some common interactions:
+
+* Any `l::AbstractLayer` has a method for `Functors.functor`, 
+  thus you need not invoke `@functor`. Note that your `struct` should
+  have the default constructor. (Or something which similarly accepts all of
+  its fields as arguments. It is simplest not to write an inner constructor.)
+
+* Calling `Adapt.adapt` on any `l::AbstractLayer` will recurse using `Functors.fmap`,
+  ensuring that the identification between tied weights is perserved.
+
+* Subtypeing `PartialTrainLayer` marks only some fields as trainable,
+  by overloading `Optimisers.trainable`.
+
+* Some subtypes tell fancy `show` whether to unfold their contents: 
+  `l::ContainerLayer` behaves like `Chain`, while `l::SimpleLayer` behaves like `Dense`.
+"""
+abstract type AbstractLayer end
+
+function Functors.functor(::Type{T}, x) where {T<:AbstractLayer}
+  if @generated
+    F = fieldnames(T)
+    args = map(sy -> :(getfield(x, $(QuoteNode(sy)))), F)
+    C = Base.typename(T).name  # constructor
+    recon = VERSION > v"1.9-" ? :(Splat($C)) : :(Base.splat($C))
+    :((NamedTuple{$F}(($(args...),)), $recon))
+  else
+    # Getting this parameterless type takes about 2μs, every time:
+    namedtuple(x), Base.splat(Base.typename(T).wrapper)
+  end
+end
+
+function namedtuple(x::T) where T
+  F = fieldnames(T)
+  NamedTuple{F}(map(sy -> getfield(x, sy), F))
+end
+
+Adapt.adapt_structure(to, layer::AbstractLayer) = fmap(x -> adapt(to, x), layer)
```

Comment:
Can be just this
```suggestion
Adapt.adapt_structure(to, layer::AbstractLayer) = fmap(adapt(to), layer)
```
but check the version bound on Adapt